src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,143 @@ public static Path of(URI uri) {
+     *
+     * @param   ext
+     *          the first extension for which to check
+     *
+     * @param   extensions
+     *          zero or more subsequent extensions for which to check
+     *
+     * @return  an {@code Optional} which either contains the found extension
+     *          or is {@link Optional#empty empty} if this path either does not
+     *          have an extension or its extension is not among those given
+     *
+     * @since 19
+     */
+    default Optional<String> hasExtension(String ext, String... extensions) {
+        String thisExtension = getExtension().orElse(null);
+        if (thisExtension == null)
+            return Optional.empty();
+
+        Objects.requireNonNull(ext);
+        if (ext.equals(thisExtension))
+            return Optional.of(ext);
+
+        for (String e : extensions) {
+            Objects.requireNonNull(e);
+            if (e.equals(thisExtension))
+                return Optional.of(e);
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Returns a {@code Path} which is equal to this path aside from its
+     * extension which is replaced by the provided extension. If the provided
+     * extension is {@link String#isEmpty() empty}, then the current extension
+     * and the dot ('.') preceding it are removed. If this path does not have
+     * an extension then the provided extension is appended.
+     *
+     * @implNote
+     * This default implementation works with the string representation of this
+     * path. Paths which have a native representation should override this
+     * implementation in order to preserve that representation.
+     *
+     * @param  extension
+     *         the extension to replace this path's current extension;
+     *         if non-empty, may not contain a leading or trailing dot
+     *         but is not otherwise checked for validity
+     *
+     * @return a path equal to this path but with extension replaced by
+     *         that provided
+     *
+     * @throws IllegalArgumentException if the provided extension contains a
+     *         leading or trailing dot ('.')
+     *
+     * @since 19
+     */
+    default Path replaceExtension(String extension) {
+        Objects.requireNonNull(extension, "extension");
+
+        // verify the extension contains neither a leading nor trailing dot
+        if (!extension.isEmpty()) {
+            if (extension.charAt(0) == '.' ||
+                extension.charAt(extension.length() - 1) == '.')
+                throw new IllegalArgumentException("leading or trailing dot");
+        }
+
+        String thisPath = toString();
+        String thisExtension = getExtension().orElse(null);
+        FileSystem fs = this.getFileSystem();
+
+        // if this path has no extension, append that provided
+        if (thisExtension == null)
+            return extension.isEmpty() ?
+                this : fs.getPath(thisPath + "." + extension);
```

Comment:
Can this end up with a double dot ".." in the case where the original path ended in in dot?
I.e.
```
    Path p = Path.of("foo.");
    Path q = p.replaceExtension("bar");
```

Reply:
Unfortunately the result of the above is currently `foo..bar`. If a Path which ends in a dot is considered to have `""` as an extension instead of `Optional.empty()`, that might simplify this case. Needs further scrutiny.